README.md
# https-pool

A cached pool of https server for domains.

---

[![NPM version](https://img.shields.io/npm/v/https-pool.svg?style=flat)](https://npmjs.org/package/https-pool)
[![Build Status](https://img.shields.io/travis/ahungrynoob/https-pool.svg?style=flat)](https://travis-ci.org/ahungrynoob/https-pool)
[![codecov.io](https://img.shields.io/codecov/c/github/ahungrynoob/https-pool.svg?style=flat)](http://codecov.io/github/ahungrynoob/https-pool?branch=master)
[![NPM downloads](http://img.shields.io/npm/dm/https-pool.svg?style=flat)](https://npmjs.org/package/https-pool)

## Install

```bash
$ npm i https-pool --save
```

## Usage

> https-pool is useful for getting https server without worry about forging certificates.

Just pass your CA certificate options into HttpsPool, and get a empty https server pool.

```javascript
const HttpsPool = require("https-pool");
const fs = require("fs");

const key = fs.readFileSync(path.join(fixtures, "root.key"));
const cert = fs.readFileSync(path.join(fixtures, "root.crt"));

const httpsPool = new HttpsPool({
  key,
  cert,
  commonName: "example",
  countryName: "CN",
  ST: "SH",
  localityName: "SH",
  organizationName: "example.com",
  OU: "example.com"
});

// You can save the cert and key of root CA to let the client trust it.
const { key, cert } = httpsPool.CA;
```

You can get a https server like this ↓. That's enough for common usage. `https-pool` will take care of the cache logic, so worrying about your memory is unnecessary.

```javascript
httpsPool.getServer(
  "www.foo.com",
  (req, res) => {
    // the listener for server.request event
    console.log(req.headers);
    res.send("ok");
  },
  _port => {
    // callback with a random available port
    console.log(typeof _port === "number");
  },
  // timeout for the new https-server
  3000
);
```

## API

### Properties

- `httpsPool.CA` - return as {key, cert} in the format as pem. (You could trust the key and cert on the your client such as browser)

### HttpsPool(options)

It will create a https pool.

- options
  - `option.timeout` - https server won't close until secure conntection is established within timeout (default 6000)
  - `option.max_servers` - max num for https servers the pool cached (default 220)
  - `option.key` and `option.cert`- the private key and cert of the root ca which your client trusts or https-pool will create one
  - If you don't pass `option.key` and `option.cert`, you need to pass these options:
    - `option.commonName` - the common name option
    - `option.countryName` - the country name option
    - `option.ST` - the ST option
    - `option.localityName` - the locality name option
    - `option.organizationName` - the organization name option
    - `option.OU` - the OU option

### HttpsPool#getServer(hostname, listener, callback, timeout)

It will return a https server if available or will create one and cached.

- `hostname` - hostname which https server base on
- `listener` {`Function` | `Object`} - request event listener or Object type with custom event listener
- `callback` - callback func with port arg
- `timeout` - timeout for https server in ms

### HttpsPool#existsServer(hostname)

Tell whether the server basing on the hostname exists

- `hostname` - the hostname server base on
- `return {boolean}`

### HttpsPool#removeServer(hostname)

Remove the server in the cache

- `hostname` - the hostname server base on

### HttpsPool#free()

Free the https server who has no connections when count > max

### HttpsPool#clear()

Clear the https pool forcily

## LICENSE

Licensed under the MIT license.


CHANGELOG.md
<a name="1.0.1"></a>
## [1.0.1](https://github.com/ahungrynoob/https-pool/compare/v1.0.0...v1.0.1) (2019-10-09)


### Bug Fixes

* enable specific key and cert for ca ([#4](https://github.com/ahungrynoob/https-pool/issues/4)) ([3a27bc4](https://github.com/ahungrynoob/https-pool/commit/3a27bc4))



<a name="1.0.0"></a>
# [1.0.0](https://github.com/ahungrynoob/https-pool/compare/d0d52be...v1.0.0) (2019-10-09)


### Features

* https-pool ([#3](https://github.com/ahungrynoob/https-pool/issues/3)) ([d0d52be](https://github.com/ahungrynoob/https-pool/commit/d0d52be))





test/index.test.js
'use strict';

const assert = require('assert');
const mm = require('mm');
const net = require('net');
const request = require('request');
const path = require('path');
const fs = require('fs');
const HttpsPool = require('../index');

const fixtures = path.join(__dirname, 'fixtures');

describe('lib/https-pool.js', () => {
  let httpsPool;
  beforeEach(() => {
    const key = fs.readFileSync(path.join(fixtures, 'root.key'));
    const cert = fs.readFileSync(path.join(fixtures, 'root.crt'));
    httpsPool = new HttpsPool({
      key,
      cert,
      commonName: 'example',
      countryName: 'CN',
      ST: 'SH',
      localityName: 'SH',
      organizationName: 'example.com',
      OU: 'example.com',
    });
  });

  afterEach(() => {
    httpsPool.clear();
  });

  describe('httpsPool#constructor', () => {
    it('should return a https pool instance and get default options', () => {
      assert(httpsPool.TIMEOUT === 6000);
      assert(httpsPool.MAX_SERVERS === 220);
      assert(typeof httpsPool.CA.cert === 'string' && typeof httpsPool.CA.key === 'string');
    });

    it('should return the same private key in pem format', () => {
      const key = fs.readFileSync(path.join(fixtures, 'root.key'), 'utf8');
      assert.strictEqual(key.replace(/[\r\n]/g, ''), httpsPool.CA.key.replace(/[\r\n]/g, ''));
    });
  });

  describe('httpsPool#getServer', () => {
    it('should get cached https server', done => {
      httpsPool.getServer('www.foo.com');
      httpsPool.getServer('www.foo.com', null, port => {
        assert(typeof port === 'number');
        done();
      });
    });

    it('should emit request event when listener is function', done => {
      let port;
      httpsPool.getServer('www.foo.com', req => {
        assert.strictEqual(req.headers.host, '127.0.0.1:' + port);
        done();
      }, _port => {
        port = _port;
        request.get({
          url: 'https://127.0.0.1:' + port,
          rejectUnauthorized: false,
        });
      });
    });

    it('should emit request event when listener is object', done => {
      let port;
      httpsPool.getServer('www.foo.com', {
        request: req => {
          assert.strictEqual(req.headers.host, '127.0.0.1:' + port);
          done();
        },
      }, _port => {
        port = _port;
        request.get({
          url: 'https://127.0.0.1:' + port,
          rejectUnauthorized: false,
        });
      });
    });

    it('should remove server when emit error', done => {
      httpsPool.getServer('www.foo.com', null, () => {
        httpsPool._cache['www.foo.com'].server.emit('error');
        assert(httpsPool._cache['www.foo.com'] === undefined);
        done();
      });
    });

    it('should remove server when emit tlsClientError', done => {
      httpsPool.getServer('www.foo.com', null, () => {
        const mockConn = { destroy: () => {} };
        httpsPool._cache['www.foo.com'].server.emit('tlsClientError', new Error('mock err'), mockConn);
        assert(httpsPool._cache['www.foo.com'] === undefined);
        done();
      });
    });

    it('should remove server when no connection happens with timeout', done => {
      httpsPool = new HttpsPool({
        timeout: 1000,
        commonName: 'example',
        countryName: 'CN',
        ST: 'SH',
        localityName: 'SH',
        organizationName: 'example.com',
        OU: 'example.com',
      });

      httpsPool.getServer('www.foo.com');
      setTimeout(() => {
        assert(httpsPool._cache['www.foo.com'] === undefined);
        done();
      }, 1500);
    });
  });

  describe('httpsPool#existsServer', () => {
    it('should return true when pool has server', done => {
      httpsPool.getServer('www.foo.com', null, () => {
        assert.strictEqual(httpsPool.existsServer('www.foo.com'), true);
        done();
      });
    });

    it('should return false when pool has no such server', done => {
      httpsPool.getServer('www.foo.com', null, () => {
        assert.strictEqual(httpsPool.existsServer('www.example.com'), false);
        done();
      });
    });
  });

  describe('httpsPool#removeServer', () => {
    it('should return undefined after remove server', done => {
      httpsPool.getServer('www.foo.com', null, () => {
        httpsPool.removeServer('www.foo.com');
        assert(httpsPool._cache['www.foo.com'] === undefined);
        done();
      }, 10000);
    });

    it('should return undefined after remove inexistent server', () => {
      httpsPool.removeServer('www.foo.com');
      assert(httpsPool._cache['www.foo.com'] === undefined);
    });
  });

  describe('httpsPool#free', () => {
    it('should get correct server count', done => {
      httpsPool = new HttpsPool({
        max_servers: 2,
        commonName: 'example',
        countryName: 'CN',
        ST: 'SH',
        localityName: 'SH',
        organizationName: 'example.com',
        OU: 'example.com',
      });
      assert(httpsPool.MAX_SERVERS === 2);
      httpsPool.getServer('www.foo.com', null, () => {
        httpsPool.getServer('www.bar.com', null, () => {
          httpsPool.free();
          const serverCount = Object.keys(httpsPool._cache).length;
          assert(serverCount === 1);
          done();
        });
      });
    });

    it('should get correct server count when secure connection established', done => {
      httpsPool = new HttpsPool({
        max_servers: 2,
        commonName: 'example',
        countryName: 'CN',
        ST: 'SH',
        localityName: 'SH',
        organizationName: 'example.com',
        OU: 'example.com',
      });
      mm(net.Server.prototype, 'getConnections', function(cb) {
        return cb(null, 1);
      });

      httpsPool.getServer('www.foo.com', null, () => {
        httpsPool.getServer('www.bar.com', null, () => {
          httpsPool.free();
          const serverCount = Object.keys(httpsPool._cache).length;
          assert(serverCount === 2);
          done();
        });
      });
    });
  });

  describe('httpsPool#clear', () => {
    it('should return empty cache after clear', () => {
      httpsPool.clear();
      assert.deepEqual(httpsPool._cache, {});
    });
  });
});


test/lib/utils.test.js
'use strict';

const mm = require('mm');
const path = require('path');
const forge = require('node-forge');
const fs = require('fs');
const assert = require('assert');
const { createCACert, createCertificate, defer } = require('../../lib/utils');

const pki = forge.pki;

const fixtures = path.join(__dirname, '../fixtures');

// const certificate = createCertificate('www.taobao.com', ca.key, ca.cert);

describe('lib/utils', () => {
  afterEach(mm.restore);
  it('should generate correct root ca pem without option.key', () => {
    mm(forge.pki.rsa, 'generateKeyPair', function() {
      const privateKeyPem = fs.readFileSync(path.join(fixtures, 'root.key'));
      const privateKey = pki.privateKeyFromPem(privateKeyPem);
      const publicKey = pki.setRsaPublicKey(privateKey.n, privateKey.e);
      return { privateKey, publicKey };
    });

    const RealDate = Date;

    mm(global, 'Date', class extends RealDate {
      constructor() {
        return new RealDate(1559742360000);
      }

      getFullYear() {
        return 2019;
      }
    });

    const ca = createCACert({
      commonName: 'whistle.1559742360747607',
      countryName: 'CN',
      ST: 'ZJ',
      localityName: 'HZ',
      organizationName: '1559742360747607.wproxy.org',
      OU: 'wproxy.org',
    });

    const caPem = pki.certificateToPem(ca.cert);
    const rootCrt = fs.readFileSync(path.join(fixtures, 'root.crt'), 'utf8');
    assert.strictEqual(caPem.replace(/[\r\n]/g, ''), rootCrt.replace(/[\r\n]/g, ''));
  });

  it('should generate corresponding root ca pem with option.key', () => {
    const key = fs.readFileSync(path.join(fixtures, 'root.key'));
    const RealDate = Date;

    mm(global, 'Date', class extends RealDate {
      constructor() {
        return new RealDate(1559742360000);
      }

      getFullYear() {
        return 2019;
      }
    });

    const ca = createCACert({
      key,
      commonName: 'whistle.1559742360747607',
      countryName: 'CN',
      ST: 'ZJ',
      localityName: 'HZ',
      organizationName: '1559742360747607.wproxy.org',
      OU: 'wproxy.org',
    });

    const caPem = pki.certificateToPem(ca.cert);
    const rootCrt = fs.readFileSync(path.join(fixtures, 'root.crt'), 'utf8');
    assert.strictEqual(caPem.replace(/[\r\n]/g, ''), rootCrt.replace(/[\r\n]/g, ''));
  });

  it('should generate correct certificate based on hostname', () => {
    const privateKeyPem = fs.readFileSync(path.join(fixtures, 'root.key'));
    const privateKey = pki.privateKeyFromPem(privateKeyPem);
    const certPem = fs.readFileSync(path.join(fixtures, 'root.crt'));
    const cert = pki.certificateFromPem(certPem);

    const certificateDomain = createCertificate('www.taobao.com', privateKey, cert);
    assert.strictEqual(
      certificateDomain.cert.subject.attributes[0].value,
      'www.taobao.com'
    );

    const certificateIP = createCertificate('192.168.1.1', privateKey, cert);
    assert.strictEqual(
      certificateIP.cert.subject.attributes[0].value,
      '192.168.1.1'
    );
  });

  it('should trigger then callback when call defer.resolve', done => {
    const { promise, resolve } = defer();
    promise.then(value => {
      assert.strictEqual(value, 'foo');
      done();
    });
    setTimeout(() => {
      resolve('foo');
    }, 2000);
  });

  it('should trigger catch callback when call defer.reject', done => {
    const { promise, reject } = defer();
    promise.catch(value => {
      assert.strictEqual(value, 'bar');
      done();
    });
    setTimeout(() => {
      reject('bar');
    }, 2000);
  });
});


lib/utils.js
'use strict';

const forge = require('node-forge');
const crypto = require('crypto');
const net = require('net');

const pki = forge.pki;

/**
 * create certificate
 * @param {pki.rsa.PublicKey} publicKey - the public key for a cert
 * @param {string} serialNumber - serial number for a cert
 * @default '01'
 */
function createCert(publicKey, serialNumber = '01') {
  const cert = pki.createCertificate();
  cert.publicKey = publicKey;
  cert.serialNumber = serialNumber;
  const curYear = new Date().getFullYear();
  cert.validity.notBefore = new Date();
  cert.validity.notAfter = new Date();
  cert.validity.notBefore.setFullYear(curYear - 1);
  cert.validity.notAfter.setFullYear(curYear + 10);
  return cert;
}

/**
 * create a root CA cert with privateKey pem or will use own privateKey
 *
 * @param {Object} option - the subect and issuer options for the root ca
 * @param {string} option.commonName - the common name option
 * @param {string} option.countryName - the country name option
 * @param {string} option.ST - the ST option
 * @param {string} option.localityName - the locality name option
 * @param {string} option.organizationName - the organization name option
 * @param {string} option.OU - the OU option
 * @return {Object} - return private key and cert
 */
function createCACert(option) {
  const { commonName, countryName, ST, localityName, organizationName, OU } = option;
  const { privateKey, publicKey } = pki.rsa.generateKeyPair(2048);

  const cert = createCert(publicKey);
  const attrs = [{
    name: 'commonName',
    value: commonName,
  }, {
    name: 'countryName',
    value: countryName,
  }, {
    shortName: 'ST',
    value: ST,
  }, {
    name: 'localityName',
    value: localityName,
  }, {
    name: 'organizationName',
    value: organizationName,
  }, {
    shortName: 'OU',
    value: OU,
  }];

  cert.setSubject(attrs);
  cert.setIssuer(attrs);
  cert.setExtensions([{
    name: 'basicConstraints',
    cA: true,
  }, {
    name: 'keyUsage',
    keyCertSign: true,
    digitalSignature: true,
    nonRepudiation: true,
    keyEncipherment: true,
    dataEncipherment: true,
  }, {
    name: 'extKeyUsage',
    serverAuth: true,
    clientAuth: true,
    codeSigning: true,
    emailProtection: true,
    timeStamping: true,
  }, {
    name: 'nsCertType',
    client: true,
    server: true,
    email: true,
    objsign: true,
    sslCA: true,
    emailCA: true,
    objCA: true,
  }]);

  cert.sign(privateKey, forge.md.sha256.create());

  return {
    key: privateKey,
    cert,
  };
}

/**
 * create certificate for a specific hostname
 * @param {string} hostname - the hostname which certificate based on
 * @param {pki.rsa.PrivateKey} privateKey - your private key of the root ca
 * @param {pki.Certificate} ca - the root ca
 * @return {Object} - return the private key and cert in pem format
 */
function createCertificate(hostname, privateKey, ca) {
  const RANDOM_SERIAL = '.' + Date.now() + '.' + Math.floor(Math.random() * 10000);
  const serialNumber = crypto.createHash('sha1')
    .update(hostname + RANDOM_SERIAL, 'binary').digest('hex');
  const cert = createCert(ca.publicKey, serialNumber);

  cert.setSubject([{
    name: 'commonName',
    value: hostname,
  }]);

  cert.setIssuer(ca.subject.attributes);
  cert.setExtensions([{
    name: 'subjectAltName',
    altNames: [ net.isIP(hostname) ?
      {
        type: 7,
        ip: hostname,
      } : {
        type: 2,
        value: hostname,
      } ],
  }]);
  cert.sign(privateKey, forge.md.sha256.create());

  return {
    key: privateKey,
    cert,
  };
}

function defer() {
  let resolve;
  let reject;
  const promise = new Promise((res, rej) => {
    resolve = res;
    reject = rej;
  });
  return {
    promise,
    resolve,
    reject,
  };
}

module.exports = {
  createCACert,
  createCertificate,
  defer,
};


lib/https-pool.js
'use strict';

const https = require('https');
const forge = require('node-forge');
const detect = require('detect-port');
const { createCertificate, createCACert, defer } = require('./utils');

const pki = forge.pki;
const LOCALHOST = '127.0.0.1';

class HttpsPool {
  /**
   * https pool to get certificated https server
   * @param {Object} option - option for https pool
   * @param {number} option.timeout - https server won't close until secure conntection is established within timeout
   * @default 6000
   * @param {number} option.max_servers - max num for https servers the pool cached
   * @default 220
   * the subect and issuer options for the root ca
   * @param {string} option.key - the private key for the root ca in pem format
   * @param {string} option.cert - the cert for the root ca in pem format
   * @param {string} option.commonName - the common name option
   * @param {string} option.countryName - the country name option
   * @param {string} option.ST - the ST option
   * @param {string} option.localityName - the locality name option
   * @param {string} option.organizationName - the organization name option
   * @param {string} option.OU - the OU option
   */
  constructor({
    timeout,
    max_servers,
    key,
    cert,
    commonName,
    countryName,
    ST,
    localityName,
    organizationName,
    OU,
  }) {
    this._cache = {};
    this.TIMEOUT = timeout || 6000;
    this.MAX_SERVERS = max_servers || 220;
    if (key && cert) {
      this._rawCA = {
        key: pki.privateKeyFromPem(key),
        cert: pki.certificateFromPem(cert),
      };
    } else {
      this._rawCA = createCACert({
        commonName,
        countryName,
        ST,
        localityName,
        organizationName,
        OU,
      });
    }
    this.CA = {
      key: pki.privateKeyToPem(this._rawCA.key),
      cert: pki.certificateToPem(this._rawCA.cert),
    };
  }

  /**
   * it will return a https server if available or will create one with cert
   * @param {string} hostname - hostname which https server base on
   * @param {Function | Object} listener - request event listener or Object type with custom event listener
   * @param {Function} callback - callback func with port arg
   * @param {number} timeout - timeout for https server in ms
   */
  getServer(hostname, listener, callback, timeout) {
    if (this._cache[hostname]) {
      this._cache[hostname].then(callback);
      return;
    }

    const { promise, resolve } = defer();
    this._cache[hostname] = promise;
    this._cache[hostname].then(callback);

    this.free();

    // start create server
    const option = createCertificate(hostname, this._rawCA.key, this._rawCA.cert);
    option.key = pki.privateKeyToPem(option.key);
    option.cert = pki.certificateToPem(option.cert);

    const server = https.createServer(option);

    if (typeof listener === 'function') {
      server.on('request', listener);
    } else if (listener) {
      for (const event in listener) {
        server.on(event, listener[event]);
      }
    }

    if (/^\d+$/.test(timeout)) {
      server.timeout = timeout;
    }

    // error handle
    server.on('error', () => this.removeServer(hostname));
    server.once('tlsClientError', () => {
      this.removeServer(hostname);
    });

    detect().then(port => {
      server.listen(port, LOCALHOST, () => {
        const timer = setTimeout(() => this.removeServer(hostname), this.TIMEOUT);
        server.once('secureConnection', function() {
          clearTimeout(timer);
        });
        promise.server = server;
        resolve(port);
      });
    }).catch(
      /* istanbul ignore next */
      () => {}
    );
  }

  /**
   * whether the server basing on the hostname exists
   * @param {string} hostname - the hostname server base on
   * @return {boolean} -
   */
  existsServer(hostname) {
    const promise = this._cache[hostname];
    if (promise) {
      return true;
    }
    return false;
  }

  /**
   * remove the server in the cache
   * @param {string} hostname - the hostname server base on
   */
  removeServer(hostname) {
    const promise = this._cache[hostname];
    if (!promise) {
      return;
    }
    delete this._cache[hostname];

    try {
      promise.server && promise.server.close();
    } catch (e) {
      // placeholder
    }
  }

  /**
   * free the https server without connections when count > max
   */
  free() {
    /*
    * TODO: use lock in the future, We use Object.keys() now for multi process.
    * [o(n) complexity]
    */
    const server_count = Object.keys(this._cache).length;
    if (server_count < this.MAX_SERVERS) {
      return;
    }
    for (const hostname in this._cache) {
      const promise = this._cache[hostname];
      const server = promise.server;
      if (promise && server && !promise._pending) {
        promise._pending = true;
        server.getConnections((err, count) => {
          promise._pending = false;
          if (!err && !count && server_count >= this.MAX_SERVERS) {
            this.removeServer(hostname);
          }
        });
      }
    }
  }

  /**
   * clear the https pool forcily
   */
  clear() {
    for (const hostname in this._cache) {
      this.removeServer(hostname);
    }
    this._cache = {};
  }
}

module.exports = HttpsPool;
